### DISCUSSION TITLE
Grouping change description

### DISCUSSION THREAD

**Discussion by bwgjoseph** [javers/javers#1128] [Q&A] [answered]
Hi,

While looking through [type-of-changes](https://javers.org/documentation/diff-examples/) and trying out, it seems like every change can be pulled out by getting `diff.getChangesByType(ValueChange.class)`, or at least based on my sample data set. 

So my objective is to output a user-friendly change description based on the changes between the two object diff.

For example, given the following data structure

```java
class Person implements Serializable {
    @Id
    private int id;
    private String firstName;
    private String lastName;
    private List<Address> addresses;
    private List<String> hobbies;
    private Gender gender;
    private boolean isEmployed;
    private CreditCard creditCard;
    private LocalDateTime lastUpdated;
}

class Address implements Serializable {
    private String street;
    private String postalCode;
    private String block;
}

class CreditCard implements Serializable {
    private long cardNumber;
    private int cvv;
    private LocalDate expiry;
}
```

And I have something along this line

```java
d.getChangesByType(ValueChange.class)
    .forEach(vc -> {
        if (vc.getLeft() != null) {
            log.info("{} changed from {} to {}", vc.getPropertyName(), vc.getLeft(), vc.getRight());
        } else {
            log.info("{} is {}", vc.getPropertyName(), vc.getRight());
        }
    });
```

My output is something like

```
// I added a new address to the list
street is Cesar Extensions
postalCode is 47795
block is 056
// I changed creditcard info
cardNumber changed from 4240975573591258006 to -585401770101026848
cvv changed from 1146085061 to -2006036587
// I changed the gender
gender changed from MALE to FEMALE
// I changed isEmployed flag
isEmployed changed from true to false
```

And so my question is if it is possible to identify and group the information based on the data type? So it would look something like this

```
// knowing there's a new address added to the List<Address>
New address added:
  - street is Cesar Extensions, postalCode is 47795, block is 056
// if removed, it can be
Existing address removed:
  - street is Cesar Extensions, postalCode is 47795, block is 056
// CreditCard is an Object (Value Object?)
Existing CreditCard Info updated:
  - cardNumber changed from 4240975573591258006 to -585401770101026848, cvv changed from 1146085061 to -2006036587
// and the rest of the primitive data type
gender changed from MALE to FEMALE
isEmployed changed from true to false
```

I'm not exactly sure how to detect for added/removed data using `getChangesByType`, as calling `getChangeType()` will give me `PROPERTY_VALUE_CHANGED` no matter added or removed. I did try to use `groupByObject()` but I couldn't quite seem to get it to be what I wanted. 

Is there any way to achieve it? 

Thanks!

---

Update: 

I just realize that changes to `List<String>` or rather `List<PrimitiveDataType>` would be listed in `ListChange` and not `ValueChange`, and I think it might be related to https://github.com/javers/javers/issues/1094 too

**Comment by bartoszwalacik** [comment]
As you can see in the Changes class API , there are two methods for grouping : `groupByCommit()` and `groupByObject()`.

**Reply by bwgjoseph** [reply]
Hi @bartoszwalacik,

Apologies for the late response.

I'm using it without commits, basically just comparing between two objects. Hence, if I'm not wrong, I won't be able to use `groupByCommits`? As for `groupByObject`, it doesn't seem to have a way to get the before and after value. Please correct me if I'm wrong.

For example, when I added new item to the address list, and I print out the info like such

```java
d.groupByObject().forEach(obj -> {
    obj.getPropertyChanges().forEach(changes -> {
        log.info("start");
        log.info("each getAffectedGlobalId {}", changes.getAffectedGlobalId());
        log.info("each getAffectedLocalId {}", changes.getAffectedLocalId());
        log.info("each getAffectedObject {}", changes.getAffectedObject());
        log.info("each getChangeType {}", changes.getChangeType());
        log.info("each getClass {}", changes.getClass());
        log.info("each getCommitMetadata {}", changes.getCommitMetadata());
        log.info("each getPropertyName {}", changes.getPropertyName());
        log.info("each getPropertyNameWithPath {}", changes.getPropertyNameWithPath());
        log.info("end");
    });
});
```

And this is what I get back

```log
start
each getAffectedGlobalId Person/1#addresses/2
each getAffectedLocalId null
each getAffectedObject Optional[Address(street=Mark Trail, postalCode=08368, block=8928)]
each getChangeType PROPERTY_VALUE_CHANGED
each getClass class org.javers.core.diff.changetype.InitialValueChange
each getCommitMetadata Optional.empty
each getPropertyName street
each getPropertyNameWithPath addresses/2.street
end
```

And that repeats for 3 times which I think because there's 3 fields in it. So if I have a huge number of fields, this would be repeated for a x number of times.

I was wondering if there is any existing API that I missed, where I can grab by the type such as:

- Item Removed
- Item Added
- Item Changed

So for the `Item Changed` in this case, would report a `List<Address>` changed, and new `address` is added. Or if there is something along this line? I've been trying with the different APIs but doesn't seem to have the ones I wanted. So I'm not sure if I'm missing something obvious or is it that the library does not have such a API.

If it does have, do direct me to some example, or if not, would appreciate your help with some samples

Thanks!

**Comment by bartoszwalacik** [comment, marked as answer]
There is no such API, I mean for selecting subtypes of CollectinChange. If you can suggest some improvements to Change API we can talk about it.
